packages/mui-material/src/TablePagination/TablePagination.test.js

```diff
@@ -706,4 +706,28 @@ describe('<TablePagination />', () => {
 
     expect(combobox.parentElement).not.to.have.attribute('variant');
   });
+
+  it('should not raise a warning for hiddenLabel when the variant is filled', () => {
```

Comment:
We don't need this test. Instead, we should add a TypeScript test in `packages\mui-material\src\TablePagination\TablePagination.spec.tsx` to test that hiddenLabel prop is available when filled variant is used. Basically, TypeScript shouldn't throw a compile error as mentioned in #38487.

Reply:
@ZeeshanTamboli I have tried out a plethora of different solution to solve the type issue inside of Select.d.ts, however the complexity of the issue is very high. I have found no good way of extending the interface with the specific proptype depending on which variant is selected. I believe the main issue lies in that in typescript an interface can only be extended by object types with statically known members, so we cannot perform a similar approach to what I did in the affermentioned solution. 

Another issue lies in that since we default the variant value to "outlined", Typescript will use this as the variant when checking which inputProps it should contain if we conditionally extend with them. Thus when we pass a variant value at runtime TypeScript's static type checking won't automatically adjust to the runtime value, so the typeError will still appear. 

A way the type error could be removed is by extending the interface with all variations of inputProps, however in my opinion I believe solving the issue for the specific components that use the Select component is a good alternative.

Reply:
@ZeeshanTamboli I have tried out a plethora of different solution to solve the type issue inside of Select.d.ts, however the complexity of the issue is very high. I have found no good way of extending the interface with the specific proptype depending on which variant is selected. I believe the main issue lies in that in typescript an interface can only be extended by object types with statically known members, so we cannot perform a similar approach to what I did in the affermentioned solution.

Another issue lies in that since we default the variant value to "outlined", Typescript will use this as the variant when checking which inputProps it should contain if we conditionally extend with them. Thus when we pass a variant value at runtime TypeScript's static type checking won't automatically adjust to the runtime value, so the typeError will still appear. If we do not give a default variant however, this would require us to have Variant as an input to Select, which would be a breaking change for the component. 

A way the type error could be removed is by extending the interface with all variations of inputProps, however in my opinion I believe solving the issue for the specific components that use the Select component is a good alternative.

Reply:
I tried with the following change:
```diff
--- a/packages/mui-material/src/Select/Select.d.ts
+++ b/packages/mui-material/src/Select/Select.d.ts
@@ -6,12 +6,11 @@ import { MenuProps } from '../Menu';
 import { SelectChangeEvent, SelectInputProps } from './SelectInput';
 import { SelectClasses } from './selectClasses';
 import { OutlinedInputProps } from '../OutlinedInput';
+import { FilledInputProps } from '../FilledInput';

 export { SelectChangeEvent };

-export interface SelectProps<Value = unknown>
-  extends StandardProps<InputProps, 'value' | 'onChange'>,
-    Omit<OutlinedInputProps, 'value' | 'onChange'> {
+export interface BaseSelectProps<Value> {
   /**
    * If `true`, the width of the popover will automatically be set according to the items inside the
    * menu, otherwise it will be at least the width of the select input.
@@ -144,13 +143,49 @@ export interface SelectProps<Value = unknown>
    * If the value is not an object, the string representation must match with the string representation of the option in
 order to be selected.
    */
   value?: Value | '';
+}
+
+export interface FilledSelectProps<Value>
+  extends BaseSelectProps<Value>,
+    Omit<StandardProps<FilledInputProps>, 'classes' | 'value' | 'defaultValue' | 'onChange'> {
   /**
    * The variant to use.
    * @default 'outlined'
    */
-  variant?: 'standard' | 'outlined' | 'filled';
+  variant: 'filled';
 }

+export interface StandardSelectProps<Value>
+  extends BaseSelectProps<Value>,
+    Omit<StandardProps<InputProps>, 'classes' | 'value' | 'defaultValue' | 'onChange'> {
+  /**
+   * The variant to use.
+   * @default 'outlined'
+   */
+  variant: 'standard';
+}
+
+export interface OutlinedSelectProps<Value>
+  extends BaseSelectProps<Value>,
+    Omit<StandardProps<OutlinedInputProps>, 'classes' | 'value' | 'defaultValue' | 'onChange'> {
+  /**
+   * The variant to use.
+   * @default 'outlined'
+   */
+  variant: 'outlined';
+}
+
+export type SelectVariants = 'outlined' | 'standard' | 'filled';
+
+export type SelectProps<
+  Value = unknown,
+  Variant extends SelectVariants = 'outlined',
+> = Variant extends 'filled'
+  ? FilledSelectProps<Value>
+  : Variant extends 'standard'
+  ? StandardSelectProps<Value>
+  : OutlinedSelectProps<Value>;
+
 /**
  *
  * Demos:
@@ -162,8 +197,14 @@ export interface SelectProps<Value = unknown>
  * - [Select API](https://mui.com/material-ui/api/select/)
  * - inherits [OutlinedInput API](https://mui.com/material-ui/api/outlined-input/)
  */
-declare const Select: (<Value>(props: SelectProps<Value>) => JSX.Element) & {
+export default function Select<Value = unknown, Variant extends SelectVariants = 'outlined'>(
+  props: {
+    /**
+     * The variant to use.
+     * @default 'outlined'
+     */
+    variant?: Variant;
+  } & Omit<SelectProps<Value, Variant>, 'variant'>,
+): JSX.Element & {
   muiName: string;
 };
-
-export default Select;
diff --git a/packages/mui-material/src/Select/Select.spec.tsx b/packages/mui-material/src/Select/Select.spec.tsx
index 25283d8a87..774b0da424 100644
--- a/packages/mui-material/src/Select/Select.spec.tsx
+++ b/packages/mui-material/src/Select/Select.spec.tsx
@@ -44,7 +44,7 @@ function genericValueTest() {
   <Select notched />;

   // disabledUnderline prop should be available (inherited from InputProps) and NOT throw typescript error
-  <Select disableUnderline />;
+  <Select variant="standard" disableUnderline />;

   // Tests presence of `root` class in SelectClasses
   const theme = createTheme({
```

But the `onChange` prop is causing type errors. I am not able to fix that. It would be appreciated if you could try to address this. 

If not, I'll reach out to @michaldudak to assist in resolving the issue or evaluate if the current solution is satisfactory.

Reply:
> For example if I write the test as
>
> testString(event.target.value as string);
>
> there is not error for the onChange in the test. However depending on how you write tests in MUI I am not sure if it > is ok to add type declarations in the test or not.

It would be a breaking change to do type casting. It should be able to infer properly based on `value` prop.

Reply:
I assumed so. I will continue trying to come up with a solution then.

Reply:
I made some more changes and managed to fix the issue with onChange now! Instead of extending BaseSelectProps I used intersection which solved typescript being able to infere the type on the value prop. All typescript tests pass locally except "docs:typescript". I ran yarn proptypes && yarn docs:api but it did not seem to solve the issue.

Reply:
Not sure why it fails on disableUnderline in the CI. I can see not issue when looking at it in the local spec file.
![bild](https://github.com/mui/material-ui/assets/57681017/1870b57c-0155-4a8e-9901-4dad7695e927)

Reply:
> Not sure why it fails on disableUnderline in the CI


The issue with the [failure in the CI](https://app.circleci.com/pipelines/github/mui/material-ui/115587/workflows/9d2e2acc-bc88-4b92-97dc-98d1787c1917/jobs/620662) related to `disableUnderline` in the `XGridFullDemo` seems to be due to the absence of an explicit `variant` prop in the demo file. In contrast [here](https://github.com/mui/material-ui/pull/39137#discussion_r1422168338), the component test file explicitly sets variant="standard", but this adjustment introduces a breaking change. I added it because the default variant of Select is `outlined` but now that I see that the props of Standard Input are directly supported in Select so `disableUnderline` without variant should work. 

Regarding error in `data/material/components/selects/MultipleSelectNative.tsx`, it is not able to infer the `string[]` value for the native Select.

Also, I noticed this solution actually doesn't fix #38487, because when I add:
```tsx
<TablePagination
  onPageChange={() => {}}
  page={1}
  rowsPerPage={1}
  slotProps={{
    select: {
      variant: 'filled',
    },
  }}
/>;
```
in TablePagination TypeScript test file (`TablePagination.spec.tsx`), I get the following error:
```
Type '"filled"' is not assignable to type '"outlined"'
```
which shouldn't throw.

Reply:
I made some further changes. I have removed the variant prop inside of the test, and with that I added back the props of Standard input so they are directly supported again. I removed the type error for this example

```typescript
<TablePagination
  onPageChange={() => {}}
  page={1}
  rowsPerPage={1}
  slotProps={{
    select: {
      variant: 'filled',
    },
  }}
/>;
```

By changing the SelectProps type to the following

```typescript
export type SelectProps<
  Value = unknown,
  Variant extends SelectVariants =SelectVariants,
> = BaseSelectProps<Value> &
  (Variant extends 'filled'
    ? FilledSelectProps
    : Variant extends 'standard'
    ? StandardSelectProps
    : OutlinedSelectProps);
```

The issue with data/material/components/selects/MultipleSelectNative.tsx, not being able to infer the string[] value for the native Select is still occuring though. I will try and look into it.

Reply:
@ZeeshanTamboli I have looked at it quite a bit now but haven't found any solution. Do you have any suggestion for what may cause it to occur?

Reply:
Since it's a native Select (with a `native` prop), the `onChange` callback parameter has the correct native HTML select element event type (`React.ChangeEvent<HTMLSelectElement>`). However, the demo uses `@ts-ignore` to overlook the type error where it's called. Although it's unclear why it can't infer the type based on the `value` prop (might be due to the different onChange event type?), explicitly passing the custom value type to the component resolves the issue:
```diff
--- a/docs/data/material/components/selects/MultipleSelectNative.tsx
+++ b/docs/data/material/components/selects/MultipleSelectNative.tsx
@@ -35,7 +35,7 @@ export default function MultipleSelectNative() {
         <InputLabel shrink htmlFor="select-multiple-native">
           Native
         </InputLabel>
-        <Select
+        <Select<string[]>
           multiple
           native
           value={personName}
```

Another option is to remove the `@ts-ignore` as shown below but logically it doesn't make sense to me to use `SelectChangeEvent` for the native select:
```diff
--- a/docs/data/material/components/selects/MultipleSelectNative.tsx
+++ b/docs/data/material/components/selects/MultipleSelectNative.tsx
@@ -1,7 +1,7 @@
 import * as React from 'react';
 import InputLabel from '@mui/material/InputLabel';
 import FormControl from '@mui/material/FormControl';
-import Select from '@mui/material/Select';
+import Select, { SelectChangeEvent } from '@mui/material/Select';

 const names = [
   'Oliver Hansen',
@@ -18,8 +18,8 @@ const names = [

 export default function MultipleSelectNative() {
   const [personName, setPersonName] = React.useState<string[]>([]);
-  const handleChangeMultiple = (event: React.ChangeEvent<HTMLSelectElement>) => {
-    const { options } = event.target;
+  const handleChangeMultiple = (event: SelectChangeEvent<typeof personName>) => {
+    const { options } = event.target as unknown as HTMLSelectElement;
     const value: string[] = [];
     for (let i = 0, l = options.length; i < l; i += 1) {
       if (options[i].selected) {
@@ -39,7 +39,6 @@ export default function MultipleSelectNative() {
           multiple
           native
           value={personName}
-          // @ts-ignore Typings are not considering `native`
           onChange={handleChangeMultiple}
           label="Native"
           inputProps={{
```

While option 1 seems like a good fix, there's again a consideration if it could be a breaking change for users relying on this type of native Select demo code.

Reply:
Thanks! Based on the two different version I believe the first version seems to be the better option. Even though it may be a breaking change I believe the solution is more logical than changing the event type and declaring the taget. I will update the code with the changes made in option 1.